### PR TITLE
feat: Use an explicitly closeable async iterator

### PR DESCRIPTION
**Pull request by taion**
The return method on async generators is just too broken.

### CODE REVIEW

File: src/AuthorizedSocketConnection.js [hunk 1]
```diff
@@ -237,24 +235,25 @@ export default class AuthorizedSocketConnection<TContext, TCredentials> {
...
-    this.subscriptions.forEach(async subscriptionPromise => {
-      const subscription = await subscriptionPromise;
-
-      if (subscription && typeof subscription.return === 'function') {
-        subscription.return();
-      }
-    });
+    await Promise.all([
+      this.config.credentialsManager.unauthenticate(),
+      ...Array.from(this.subscriptionContexts.values()).map(
```
Comment (jquense) [inline src/AuthorizedSocketConnection.js]: Second argument to Array.from is a map function

**Comment by taion** [comment]
Well, this is now a super awkward API.

But at least it works!

Aside from the clumsiness of the "promise for async iterator plus close function" construct, this actually... the ergonomics still aren't awful.

**Comment by taion** [comment]
The high-level idea is that we now add an explicit `close()` method to the async queues that will kill the queue. Then we clobber the `Subscriber` API to force it to return both the async iterable and a `close()` function.

We mitigate the awfulness by adding this `SubscriptionContext` object that exposes a single `close()` method that will close all subscriptions made.

**Comment by jquense** [comment]
This seems good, if not annoyingly ugly :P I guess it doesn't make sense to like try and use something like observables for all our apis and convert to a closable iterator at the graphql API edge?

**Comment by taion** [comment]
I thought about it, but actually everything aside from closing the darn thing is... really pretty nice with async iterators. The syntax is nice and they're easy to work with.

This wraps up disposing the damn things in a clean way, so... I figure, why not?

**Comment by taion** [comment]
@jquense am i good to merge this after i resolve conflicts?

File: src/AuthorizedSocketConnection.js [hunk 2]
```diff
@@ -188,8 +188,12 @@ export default class AuthorizedSocketConnection<TContext, TCredentials> {
       try {
         resultOrStream = await sourcePromise;
       } catch (err) {
-        this.subscriptionContexts.delete(id);
-        throw err;
+        if (err instanceof GraphQLError) {
```
Comment (jquense) [inline src/AuthorizedSocketConnection.js]: what a mess this is...

you know we should probably just make our own implementation of graphql :P

Comment (taion) [reply src/AuthorizedSocketConnection.js]: i should probably upstream some fix like this

Comment (jquense) [reply src/AuthorizedSocketConnection.js]: That'd only be helpful for when the merge it in two years tho

**Pull request merged**